Batch queue discussion

yaroslav-tykhonchuk opened the pull request:
Implementation of Azure batch queue. Send batches of items in a single Azure Queue message, reducing the amount of billable Send/Receive transactions.

## Sender
1. Initialize `MessageBatch<T>`, where T is a generic type parameter. You must also pass the serializer, which will convert the batch into a string to send over the network via Azure Queue. The two supported types are `JsonSerializer` and `GZipCompressedSerializer`. `GZipCompressedSerializer` compresses json serialized batch using GZip compression, which allows the batch to contain more items but takes more time and consumes more CPU via compression and decompression.
2. Use the `TryAdd()` method of a MessageBatch to fill the batch with as many items as possible. When TryAdd returns false, it means the batch is full.
3. Send created batch using `SendBatch()` method of a BatchQueue.

Example:
```csharp
var batchQueue  = new BatchQueue<string>("connectionString", "queueName", TimeSpan.FromSeconds(60));
var items = new[] { "One", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten" };
var batch = new MessageBatch<string>(SerializerType.GZipCompressed);

foreach (var item in items)
{
   var batchHasMoreSpace = batch.TryAdd(msg);

  if (batchHasMoreSpace)
      continue; // continue to add items to the batch   

  // batch is full here
  await batchQueue.SendBatch(batch);
  
  // init new batch
   batch = new MessageBatch<string>(SerializerType.GZipCompressed);
}

//Send all items that were left
if (batch.Items().Any())
    await batchQueue.SendBatch(batch);
```

## Receiver
The Receiver does not know the concept of a Batch. It receives items of a type `BatchItem<T>`, that contains the item itself (of type T) and a `Complete()` method. When all items from the batch are completed, the QueueMessage that contained this batch is deleted from the Queue.  
What happens if some batch items are not completed (stuck or failed)? They are returned to the original queue. When ? When initializing BatchQueue you need to pass `TimeSpan flushPeriod` parameter. When `flushPeriod` amount of time passes, `batchQueue` will return all not completed items back to the queue. If the batch didn't complete in `maxDequeueCount` times, it will be sent to the Quarantine queue with the name of 'quarantine-{batchQueueName}' where batchQueueName is the name of the original queue.
The Receiver allows to receive one batch (`ReceiveBatch()`) or up to 32 batches at once (`ReceiveMany`).

Example:
```csharp
var batchQueue  = new BatchQueue<string>("connectionString", "queueName", TimeSpan.FromSeconds(60));
var batchItems = await batchQueue.ReceiveBatch();
foreach (var batchItem in batchItems)
    await batchItem.Complete();
```

yaroslav-tykhonchuk committed:
Create message batch interface

mazurdmt committed:
Remove no longer used field

mazurdmt committed:
Add tests

mazurdmt committed:
Slow test

yaroslav-tykhonchuk committed:
Test batch flushing on timeout

yaroslav-tykhonchuk committed:
Move classes to separate files

yaroslav-tykhonchuk committed:
Tidy up

yaroslav-tykhonchuk committed:
Fix test

yaroslav-tykhonchuk committed:
Lease message batch while processing

yaroslav-tykhonchuk committed:
Await queue creation in test

yaroslav-tykhonchuk committed:
Decrease flush period in tests

yaroslav-tykhonchuk committed:
Dispose timer in message batch

yaroslav-tykhonchuk committed:
Ignore missing queue message on update

yaroslav-tykhonchuk committed:
Have one place to flush the message

yaroslav-tykhonchuk committed:
Throw when completing already completed message

yaroslav-tykhonchuk committed:
Rename text fixture

yaroslav-tykhonchuk committed:
Encapsulate queue client inside BatchQueue

yaroslav-tykhonchuk committed:
Set visibility timeout for queue message

yaroslav-tykhonchuk committed:
Tidy up

yaroslav-tykhonchuk committed:
Throw exception when message is too large

yaroslav-tykhonchuk committed:
Compress/decompress queue message using gzip

yaroslav-tykhonchuk committed:
Client decides whether to compress body

yaroslav-tykhonchuk committed:
Tidy up

yaroslav-tykhonchuk committed:
Trigger timer in message batch when unpacking items

yaroslav-tykhonchuk committed:
Rename

yaroslav-tykhonchuk committed:
Batch queue receives already formed batch

yaroslav-tykhonchuk committed:
Create batch with as many items as possible

yaroslav-tykhonchuk committed:
Compare compressed and uncompressed batch sizes

yaroslav-tykhonchuk committed:
BatchQueue create if not exists

yaroslav-tykhonchuk committed:
Add logger to batch queue

yaroslav-tykhonchuk committed:
Add Send/Receive queue batch sample

yaroslav-tykhonchuk committed:
Throw unexpected exception in batch queue

yaroslav-tykhonchuk committed:
Delete learning tests

yaroslav-tykhonchuk committed:
Move serialized data into separate class

yaroslav-tykhonchuk committed:
Use explicit serializers

yaroslav-tykhonchuk committed:
Pass serializer explicitly to MessageBatch

yaroslav-tykhonchuk committed:
Clean sample queue before send

yaroslav-tykhonchuk committed:
Move sample projects to Sample directory

yaroslav-tykhonchuk committed:
Add github workflow

yaroslav-tykhonchuk committed:
Add LICENSE

yaroslav-tykhonchuk committed:
Do not pack test/sample projects

yaroslav-tykhonchuk committed:
Register test adapter for the workflow

yaroslav-tykhonchuk committed:
Fix test

yaroslav-tykhonchuk committed:
Let github find workflow in default folder

yaroslav-tykhonchuk committed:
Ignore and log missing queue

yaroslav-tykhonchuk committed:
Generate random queue name in tests

yaroslav-tykhonchuk committed:
Fix test

yaroslav-tykhonchuk committed:
Fix workflow step name

yaroslav-tykhonchuk committed:
Quarantine message after maxDequeueCount attempts

Create new queue with name quarantine-{batchQueueName} and put messages there after maxDequeueCount attemplte to process a batch.

yaroslav-tykhonchuk committed:
Dequarantine messages

yaroslav-tykhonchuk committed:
Add dotnet package and nuget to Readme

yaroslav-tykhonchuk committed:
Update README.md

yaroslav-tykhonchuk committed:
Tidy up tests

yaroslav-tykhonchuk committed:
Update gitignore

yaroslav-tykhonchuk committed:
Tidy up BatchQueue

yaroslav-tykhonchuk committed:
Receive up to 32 batches at once

yaroslav-tykhonchuk committed:
Encapsulate update and delete ops in BatchQueue

yaroslav-tykhonchuk committed:
Prevent updating message that about to be quarantined

With previous implementation we had to receive message to decide whether to quarantine it or not, wasting the resources. We have this info already on flush time.

yaroslav-tykhonchuk committed:
Return message batch size

yaroslav-tykhonchuk committed:
Pass cancellation token

yaroslav-tykhonchuk committed:
BatchItem id consists of unique messageId and popReceipt

vansha commented on Samples/SendReceiveBatch/Receiver.cs:
```diff
@@ -0,0 +1,66 @@
+                }
+
+                Log($"Completed {batchItem.Item}");
+                await batchItem.Complete();
```
Do we need `async` here? According to the code, it's not needed here.

Conceptually, I see two options:
1. `async` and block on completing the last item in a batch.
   Pros: We have backpressure from the queue. E.g.,  if it's busy or unavailable, the client code will be aware of that.
   Cons: A little bit cumbersome async API for the client. We actually need to wait only once in a while (when completing the last item in a batch).
     
2. Don't block on complete. Batch completion always happens in a background timer thread.

Backpressure is usually a good thing in distributed systems. But the simplicity of sync API is tempting, too. My gut feeling says the first approach is right, but I'd still try to test the second one under load.

yaroslav-tykhonchuk replied:
I decided to go with option 2 now and will test it under load.

vansha commented on AzureBatchQueue/Utils/StringCompression.cs:
```diff
@@ -0,0 +1,63 @@
+
+namespace AzureBatchQueue.Utils;
+
+public static class StringCompression
```
nit: likely should be internal

vansha commented on Samples/SendReceiveBatch/Sender.cs:
```diff
@@ -0,0 +1,82 @@
+        Thread.Sleep(sleep);
+    }
+
+    private IEnumerable<MessageBatch<string>> CreateBatches(IReadOnlyList<string>? words)
```
What do you think about making MessageBatch an implementation detail and not exposing it publicly?
I mean, having client code would look more like this:

```csharp
queue.Send(items); // will make as many batches as needed and enqueue them.
```

tarasparfeniuk replied:
Wondering about having a bathcing-related `Sender`/`Reciever` code under the hood of BatchQueue. As a user, I'm interested in sending/reading items from the queue. But the way these items are batched seems like an implementation detail that may hidden inside. Keeping the queue interface as close to the original one (Azure QueueClient) as possible could simplify its integration

vansha commented on AzureBatchQueue/MessageBatch.cs:
```diff
@@ -0,0 +1,101 @@
+    {
+        items.Add(item);
+
+        var jsonSize = CalculateItemsSize();
```
It seems like we perform heavy computation (compression) many times here, should be very inneficient

vansha replied:
I've got the challenges with guessing a compressed size. 

What if we try our battle-tested approach - use blobs for large messages? Then, it would be up to clients which batch size to choose. The queue will use a blob if the message size is too large for a queue message. 

Should simplify logic a lot

vansha commented on AzureBatchQueue/QueueMessageBatch.cs:
```diff
@@ -0,0 +1,90 @@
+        completed = true;
+        await timer.DisposeAsync();
+
+        try
+        {
+            if (!BatchItems.Any())
+            {
+                await batchQueue.DeleteMessageAsync(options);
+                logger.LogDebug("Deleted queue message batch {Id}.", options.MessageId);
+                return;
+            }
+
+            await batchQueue.UpdateOrQuarantine(options, Serialize(), BatchItems.Count);
+        }
+        catch (Azure.RequestFailedException ex) when (ex.ErrorCode == "MessageNotFound")
+        {
+            logger.LogError(ex, "Accessing already flushed message with {messageId}.", options.MessageId);
+        }
+        catch (Azure.RequestFailedException ex) when (ex.ErrorCode == "QueueNotFound")
+        {
+            logger.LogError(ex, "Queue {queueName} not found when trying to delete message {messageId}.", batchQueue.Name(), options.MessageId);
+        }
```
What will happen on another, not handled exception?

vansha commented on AzureBatchQueue/QueueMessageBatch.cs:
```diff
@@ -0,0 +1,90 @@
+
+    private string Serialize() => new MessageBatch<T>(BatchItems.Select(x => x.Item).ToList(), options.SerializerType).Serialize();
+
+    public Task Complete(string id)
+    {
+        BatchItems.RemoveWhere(x => x.Id == id);
+        if (BatchItems.Any()) return Task.CompletedTask;
+
+        if (completed)
+            throw new MessageBatchCompletedException($"MessageBatch {options.MessageId} is already completed;");
+
+        try
+        {
+            timer.Change(TimeSpan.FromMilliseconds(1), Timeout.InfiniteTimeSpan);
+        }
+        catch (ObjectDisposedException)
+        {
+            throw new MessageBatchCompletedException($"MessageBatch {options.MessageId} is already completed;");
+        }
+
+        return Task.CompletedTask;
+    }
```
I assume, this code will be called from multiple threads. Can we please have tests emulating such a scenario.  I'm worried about not thread safe HashSet used here and access to timer, that might be changed both here and in Flush by other threads.

tarasparfeniuk replied:
Also concerned about thread safety. I'm a bit afraid of using timers here -- it's too unpredictable.

yaroslav-tykhonchuk replied:
Decided to use ConcurrentDictionary. It has very fast and a thread-safe method IsEmpty().

vansha commented on AzureBatchQueue/MessageBatch.cs:
```diff
@@ -0,0 +1,101 @@
+}
+
+[Serializable]
+public enum SerializerType
```
I think having both options is an overengineering here. I don't see how we could use a simple Json option, when a single mention could be larger than 48KB (unless we add support for blob storage).

vansha reviewed (changes requested):
🚀 💣 I'm really, really happy with this approach, would allow us to save tons of money 🤑 

Have some conceptual comments, though. Would be happy to join pair programming session to address them.

vansha committed:
Start azurite in docker container

vansha committed:
Rider: implicit private members

vansha committed:
Queue that works with large messages (via blobs)

yaroslav-tykhonchuk committed:
Update local storage start instructions

yaroslav-tykhonchuk committed:
Tidy up

yaroslav-tykhonchuk committed:
BatchQueue on top of message queue

yaroslav-tykhonchuk committed:
Trigger batch flush after all items complete

yaroslav-tykhonchuk committed:
Wait for timer to flush in test

Test can delete queue before Update in timer background process finishes.

yaroslav-tykhonchuk committed:
Test completing batch from multiple threads

yaroslav-tykhonchuk committed:
Remove old version of BatchQueue

yaroslav-tykhonchuk committed:
Expose azure blob port in workflow.yml

yaroslav-tykhonchuk committed:
Add default visibility timeout on receive

yaroslav-tykhonchuk committed:
Add quarantine queue

If message was unsuccessfully dequeued for maxDequeue times, BatchQueue puts the batch with not completed items into a separate queue (with a name "{queueName}-quarantine") and deletes the original message from the main queue.

yaroslav-tykhonchuk committed:
Remove redundant catch

yaroslav-tykhonchuk committed:
Update parallel test

yaroslav-tykhonchuk committed:
Log unhandled exception

yaroslav-tykhonchuk committed:
Store items in a thread-safe collection

yaroslav-tykhonchuk committed:
Throw when completing already finalized batch

yaroslav-tykhonchuk committed:
Remove legacy code

yaroslav-tykhonchuk committed:
Extract MessageQueue serializer interface

Serialize/deserialize functions only make sense when they are passed together, this grouping them into a common interface.

yaroslav-tykhonchuk committed:
Merge pull request #2 from youscan/with-blobs

BatchQueue with blobs for big messages

yaroslav-tykhonchuk committed:
BatchQueue serializer parameter

yaroslav-tykhonchuk committed:
Compress bytes to bytes

yaroslav-tykhonchuk committed:
Test updating message

yaroslav-tykhonchuk committed:
Encapsulate flush period

Send visibility timeout separately on Send and Receive, and calculate flushPeriod depending on the visibility timeout. Right now flushPeriod is just an educated (not really) guess to test TimeBatch on high load.

yaroslav-tykhonchuk committed:
Quarantine in MessageQueue

yaroslav-tykhonchuk committed:
Quarantine in BatchQueue

yaroslav-tykhonchuk committed:
Folder system in blob container

yaroslav-tykhonchuk committed:
Receive max amount of messages by default

yaroslav-tykhonchuk committed:
Make logger generic

yaroslav-tykhonchuk committed:
Update sample

yaroslav-tykhonchuk committed:
Expose batch id on batchItem (for logs)

yaroslav-tykhonchuk committed:
Fix updating large message

When message was stored in blob and need to be updated, there are 2 scenarios:
1. Message is still big, so we update the contents of the blob.
2. Message can fit into the Queue, so we delete the blob and update the contents of the QueueMessage

yaroslav-tykhonchuk committed:
Pass cancellation tokens

yaroslav-tykhonchuk committed:
Log error on quarantine

yaroslav-tykhonchuk committed:
Restrict access to internal classes

yaroslav-tykhonchuk committed:
Set completion result

To have a way to differentiate when a batch was completed by flush period or when all items were processed.

yaroslav-tykhonchuk committed:
Flush batch immediately when all complete

yaroslav-tykhonchuk committed:
Allow to init logger later

yaroslav-tykhonchuk committed:
Dequarantine messages

Move all messages from quarantine to message queue. Messages will be available as soon as they arrive to the queue.

yaroslav-tykhonchuk committed:
Pass generic logger

yaroslav-tykhonchuk committed:
Simplify calculating flush period

Calculate flush period, from the NextVisible time of each QueueMessage.

yaroslav-tykhonchuk committed:
Update response from quarantine queue

We use this method now in tests to check what actually was stored in quarantine, we don't need to create batches there.

yaroslav-tykhonchuk committed:
Make exception message general

yaroslav-tykhonchuk committed:
Expose batch item metadata

yaroslav-tykhonchuk committed:
Fix tests visibility timeout

yaroslav-tykhonchuk committed:
Fix test

yaroslav-tykhonchuk committed:
Make sure to send with zero visibility timeout

yaroslav-tykhonchuk committed:
Return result of batch item completion

yaroslav-tykhonchuk committed:
Throw exception when completing same item twice

yaroslav-tykhonchuk committed:
Do not fail on disposed timer

When we have a race condition and timer becomes disposed, the other thread may try to access timer and received ObjectDisposedException. That's why we set timer reference to null before disposing it, and make timer nullable. This is not 100% thread safe code, but want to test how it will behave.

yaroslav-tykhonchuk committed:
Synchronize threads on Timer dispose

yaroslav-tykhonchuk committed:
Do not fail when slow queue processing in tests

yaroslav-tykhonchuk committed:
Speedup internal batch collection

We don't need ConcurrentDictionary, as we already know that we can add items only in the constructor, and can only remove them during Complete phase.

yaroslav-tykhonchuk committed:
Remove unnecessary lock

yaroslav-tykhonchuk committed:
Tidy up

yaroslav-tykhonchuk committed:
Fix tests

yaroslav-tykhonchuk committed:
Throw exception on quarantine error

yaroslav-tykhonchuk committed:
Refactored payload creation

There are 2 ways to create a payload:
1. We are sending a new message, and need to create a blob if the payload size is bigger than allowed.
2. We are updating already existing message, and it can have a Blob associated with it. In case updated message contents fits into a QueueMessage - the we delete the blob, if not update the blob contents.

yaroslav-tykhonchuk committed:
Use streams instead of arrays

yaroslav-tykhonchuk committed:
Remove newtonsoft json dependency

yaroslav-tykhonchuk committed:
Test custom json compressed serializers

yaroslav-tykhonchuk committed:
Log BlobNotFound exception
